ports/stm/common-hal/sdioio/SDCard.c

```diff
@@ -0,0 +1,356 @@
+    }
+
+    HAL_SD_CardInfoTypeDef info;
+    r = HAL_SD_GetCardInfo(&self->handle, &info);
+    if (r != HAL_OK) {
+        mp_raise_ValueError_varg(translate("SDIO GetCardInfo Error %d"), (int)r);
+    }
+
+    self->num_data = 1;
+    if (num_data == 4) {
+        if ((r = HAL_SD_ConfigWideBusOperation(&self->handle, SDIO_BUS_WIDE_4B)) == HAL_SD_ERROR_NONE) {
+            DEBUG_PRINT("Switched bus to 4B mode\n");
+            self->handle.Init.BusWide = SDIO_BUS_WIDE_4B;
+            self->num_data = 4;
+        } else {
+            DEBUG_PRINT("WideBus_Enable returned %r, leaving at 1B mode\n", (int)r);
+        }
+    }
+
+    self->capacity = info.BlockNbr * (info.BlockSize / 512);
+    self->frequency = 25000000;
+
+    reserved_sdio[periph_index - 1] = true;
+
+    return;
+}
+
+uint32_t common_hal_sdioio_sdcard_get_count(sdioio_sdcard_obj_t *self) {
+    return self->capacity;
+}
+
+uint32_t common_hal_sdioio_sdcard_get_frequency(sdioio_sdcard_obj_t *self) {
+    return self->frequency; // self->frequency;
+}
+
+uint8_t common_hal_sdioio_sdcard_get_width(sdioio_sdcard_obj_t *self) {
+    return self->num_data; // self->width;
+}
+
+STATIC void check_whole_block(mp_buffer_info_t *bufinfo) {
+    if (bufinfo->len % 512) {
+        mp_raise_ValueError(translate("Buffer must be a multiple of 512 bytes"));
+    }
+}
+
+STATIC void wait_write_complete(sdioio_sdcard_obj_t *self) {
+    if (self->state_programming) {
+        HAL_SD_CardStateTypedef st = HAL_SD_CARD_PROGRAMMING;
+        // This waits up to 60s for programming to complete.  This seems like
+        // an extremely long time, but this is the timeout that micropython's
+        // implementation uses
+        for (int i=0; i < 60000 && st == HAL_SD_CARD_PROGRAMMING; i++) {
+            st = HAL_SD_GetCardState(&self->handle);
+            HAL_Delay(1);
+        };
+        self->state_programming = false;
+    }
+}
+
+STATIC void debug_print_state(sdioio_sdcard_obj_t *self, const char *what) {
+#if DEBUG_SDIO
+    HAL_SD_CardStateTypedef st = HAL_SD_GetCardState(&self->handle);
+    DEBUG_PRINT("%s, st=0x%x State=0x%x ErrorCode=0x%x\n", what, (int)st, self->handle.State, self->handle.ErrorCode);
+#endif
+}
+
+STATIC void check_for_deinit(sdioio_sdcard_obj_t *self) {
+    if (common_hal_sdioio_sdcard_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+int common_hal_sdioio_sdcard_writeblocks(sdioio_sdcard_obj_t *self, uint32_t start_block, mp_buffer_info_t *bufinfo) {
+    check_for_deinit(self);
+    check_whole_block(bufinfo);
+    wait_write_complete(self);
+    self->state_programming = true;
+    common_hal_mcu_disable_interrupts();
+    HAL_StatusTypeDef r = HAL_SD_WriteBlocks(&self->handle, bufinfo->buf, start_block, bufinfo->len / 512, 1000);
+    common_hal_mcu_enable_interrupts();
+    if (r != HAL_OK) {
+        debug_print_state(self, "after writeblocks error");
+        return -EIO;
+    }
+    // debug_print_state(self, "after writeblocks OK");
```

Comment:
I'd keep only one of these, if you want to leave them here at all.

Reply:
removed